File: server/internal/platform/setup/sampledata/inventory.go
package sampledata

import (
	"context"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"
	"github.com/z1coyan/synie/server/internal/domain/inventory/stockcount"
	"github.com/z1coyan/synie/server/internal/domain/inventory/stockdoc"
	"github.com/z1coyan/synie/server/internal/domain/inventory/stocktransfer"
	"github.com/z1coyan/synie/server/internal/platform/apierror"
	"github.com/z1coyan/synie/server/internal/platform/authz"
)

func seedOpeningStock(
	ctx context.Context, deps Dependencies, actor *authz.Actor, sc seedCtx, md masterData,
) (int, error) {
	if _, err := createStockDoc(ctx, deps, actor, sc, stockdoc.DirectionIn, sc.Warehouses.Default, 85,
		"期初建账入库(材料与通用件)", []struct {
			key string
			qty int64
		}{
			{"box_shell", 100}, {"busbar", 100}, {"mount_plate", 80}, {"terminal_assy", 80},
			{"terminal_block", 2000}, {"copper_terminal", 3000}, {"rail", 300}, {"copper_bar", 300},
			{"screw", 8000}, {"insul_sleeve", 600}, {"carton", 1000}, {"stretch_film", 100},
		}, md); err != nil {
		return 0, err
	}
	if _, err := createStockDoc(ctx, deps, actor, sc, stockdoc.DirectionIn, sc.Warehouses.Finished, 80,
		"期初建账入库(成品)", []struct {
			key string
			qty int64
		}{
			{"box_shell", 60}, {"busbar", 60}, {"mount_plate", 40}, {"terminal_assy", 40},
		}, md); err != nil {
		return 0, err
	}
	return 2, nil
}

func seedInventoryDocuments(
	ctx context.Context, deps Dependencies, actor *authz.Actor, sc seedCtx, md masterData,
) (invDocsResult, error) {
	if _, err := createStockDoc(ctx, deps, actor, sc, stockdoc.DirectionOut, sc.Warehouses.Default, 20,
		"生产领料出库", []struct {
			key string
			qty int64
		}{
			{"copper_rod", 120}, {"steel_sheet", 60}, {"copper_bar", 40}, {"screw", 1500},
		}, md); err != nil {
		return invDocsResult{}, err
	}
	if err := seedTransfer(ctx, deps, actor, sc, md); err != nil {
		return invDocsResult{}, err
	}
	if err := seedStockCount(ctx, deps, actor, sc, md); err != nil {
		return invDocsResult{}, err
	}
	return invDocsResult{StockDocs: 1}, nil
}

func createStockDoc(
	ctx context.Context, deps Dependencies, actor *authz.Actor, sc seedCtx,
	direction stockdoc.Direction, warehouseID uuid.UUID, dateAgo int, summary string,
	items []struct {
		key string
		qty int64
	},
	md masterData,
) (uuid.UUID, error) {
	docDate := daysAgo(dateAgo)
	doc, err := deps.StockDocs.Create(ctx, actor, stockdoc.CreateInput{
		Direction: direction, DocDate: &docDate, Summary: ptr(summary),
		Remarks: ptr("初始化示例库存单据"), CompanyID: sc.Company.ID, WarehouseID: warehouseID,
	})
	if err != nil {
		return uuid.Nil, err
	}
	for i, line := range items {
		mat := md.Materials[line.key]
		if _, err := deps.StockDocs.CreateItem(ctx, actor, stockdoc.CreateItemInput{
			StockDocID: doc.ID, Idx: int64(i + 1), Qty: decimal.NewFromInt(line.qty),
			MaterialID: mat.ID, UnitID: mat.DefaultUnitID,
		}); err != nil {
			return uuid.Nil, err
		}
	}
	if _, err := deps.StockDocs.Audit(ctx, actor, doc.ID); err != nil {
		return uuid.Nil, err
	}
	return doc.ID, nil
}

func seedTransfer(ctx context.Context, deps Dependencies, actor *authz.Actor, sc seedCtx, md masterData) error {
	docDate := daysAgo(10)
	transfer, err := deps.StockTransfers.Create(ctx, actor, stocktransfer.CreateInput{
		DocDate: &docDate, Summary: ptr("成品转仓调拨"), Remarks: ptr("初始化示例调拨单"),
		CompanyID: sc.Company.ID, FromWarehouseID: sc.Warehouses.Default,
		ToWarehouseID: sc.Warehouses.Finished, TransitWarehouseID: sc.Warehouses.Transit,
	})
	if err != nil {
		return err
	}
	lines := []struct {
		key string
		qty int64
	}{{"box_shell", 15}, {"terminal_block", 400}}
	itemIDs := make([]uuid.UUID, 0, len(lines))
	for i, line := range lines {
		mat := md.Materials[line.key]
		item, err := deps.StockTransfers.CreateItem(ctx, actor, stocktransfer.CreateItemInput{
			StockTransferID: transfer.ID, Idx: int64(i + 1), Qty: decimal.NewFromInt(line.qty),
			MaterialID: mat.ID, UnitID: mat.DefaultUnitID,
		})
		if err != nil {
			return err
		}
		itemIDs = append(itemIDs, item.ID)
	}
	if _, err := deps.StockTransfers.Ship(ctx, actor, transfer.ID); err != nil {
		return err
	}
	if _, err := deps.StockTransfers.Receive(ctx, actor, transfer.ID, stocktransfer.ReceiveInput{
		Receipts: []stocktransfer.Receipt{
			{ItemID: itemIDs[0], Qty: decimal.NewFromInt(15)},
			{ItemID: itemIDs[1], Qty: decimal.NewFromInt(250)},
		},
	}); err != nil {
		return err
	}
	return nil
}

func seedStockCount(ctx context.Context, deps Dependencies, actor *authz.Actor, sc seedCtx, md masterData) error {
	postingDate := daysAgo(3)
	count, err := deps.StockCounts.Create(ctx, actor, stockcount.CreateInput{
		PostingDate: &postingDate, Summary: ptr("月末例行盘点"), Remarks: ptr("初始化示例盘点单"),
		CompanyID: sc.Company.ID, WarehouseID: sc.Warehouses.Default, LoadAll: true,
	})
	if err != nil {
		return err
	}
	items, err := deps.StockCounts.ListItems(ctx, actor, count.ID)
	if err != nil {
		return err
	}
	if len(items) == 0 {
		return apierror.New(apierror.CodeConflict, "示例数据盘点整仓带出失败:默认仓库无账面余额")
	}
	screwID := md.Materials["screw"].ID
	railID := md.Materials["rail"].ID
	for _, item := range items {
		counted := item.BookQuantity
		switch item.MaterialID {
		case screwID:
			counted = item.BookQuantity.Sub(decimal.NewFromInt(50))
		case railID:
			counted = item.BookQuantity.Add(decimal.NewFromInt(5))
		}
		countedPtr := counted
		countedField := &countedPtr
		if _, err := deps.StockCounts.UpdateItem(ctx, actor, item.ID, stockcount.UpdateItemInput{
			CountedQuantity: &countedField,
		}); err != nil {
			return err
		}
	}
	if _, err := deps.StockCounts.Approve(ctx, actor, count.ID); err != nil {
		return err
	}
	return nil
}

File: server/src/modules/setup/sampledata/inventory.ts
import { decimal } from '@synie/shared'
import type { Actor } from '~/platform/authz/actor.ts'
import { ApiError } from '~/platform/http/errors.ts'
import { daysAgo, type MasterData, type SeedCtx } from './helpers.ts'
import { permitFor } from './permit.ts'
import type { SampleDataDeps } from './types.ts'

export async function seedOpeningStock(
  deps: SampleDataDeps,
  actor: Actor,
  sc: SeedCtx,
  md: MasterData,
): Promise<number> {
  await createStockDoc(
    deps,
    actor,
    sc,
    'IN',
    sc.warehouses.default,
    85,
    '期初建账入库(材料与通用件)',
    [
      { key: 'box_shell', qty: 100 },
      { key: 'busbar', qty: 100 },
      { key: 'mount_plate', qty: 80 },
      { key: 'terminal_assy', qty: 80 },
      { key: 'terminal_block', qty: 2000 },
      { key: 'copper_terminal', qty: 3000 },
      { key: 'rail', qty: 300 },
      { key: 'copper_bar', qty: 300 },
      { key: 'screw', qty: 8000 },
      { key: 'insul_sleeve', qty: 600 },
      { key: 'carton', qty: 1000 },
      { key: 'stretch_film', qty: 100 },
    ],
    md,
  )
  await createStockDoc(
    deps,
    actor,
    sc,
    'IN',
    sc.warehouses.finished,
    80,
    '期初建账入库(成品)',
    [
      { key: 'box_shell', qty: 60 },
      { key: 'busbar', qty: 60 },
      { key: 'mount_plate', qty: 40 },
      { key: 'terminal_assy', qty: 40 },
    ],
    md,
  )
  return 2
}

export async function seedInventoryDocuments(
  deps: SampleDataDeps,
  actor: Actor,
  sc: SeedCtx,
  md: MasterData,
): Promise<{ stockDocs: number }> {
  await createStockDoc(
    deps,
    actor,
    sc,
    'OUT',
    sc.warehouses.default,
    20,
    '生产领料出库',
    [
      { key: 'copper_rod', qty: 120 },
      { key: 'steel_sheet', qty: 60 },
      { key: 'copper_bar', qty: 40 },
      { key: 'screw', qty: 1500 },
    ],
    md,
  )
  await seedTransfer(deps, actor, sc, md)
  await seedStockCount(deps, actor, sc, md)
  return { stockDocs: 1 }
}

async function createStockDoc(
  deps: SampleDataDeps,
  actor: Actor,
  sc: SeedCtx,
  direction: 'IN' | 'OUT',
  warehouseId: string,
  dateAgoN: number,
  summary: string,
  items: Array<{ key: string; qty: number }>,
  md: MasterData,
): Promise<string> {
  const doc = await deps.stockDocs.create(permitFor(deps, actor, 'invStockDocs', 'create'), {
    direction,
    docDate: daysAgo(dateAgoN),
    summary,
    remarks: '初始化示例库存单据',
    companyId: sc.company.id,
    warehouseId,
  })
  for (let i = 0; i < items.length; i++) {
    const line = items[i]!
    const mat = md.materials[line.key]
    if (!mat) throw new Error(`示例物料缺失: ${line.key}`)
    await deps.stockDocs.createItem(permitFor(deps, actor, 'invStockDocItems', 'create'), {
      stockDocId: doc.id,
      idx: i + 1,
      qty: String(line.qty),
      materialId: mat.id,
      unitId: mat.defaultUnitId,
    })
  }
  await deps.stockDocs.audit(permitFor(deps, actor, 'invStockDocs', 'audit'), doc.id)
  return doc.id
}

async function seedTransfer(
  deps: SampleDataDeps,
  actor: Actor,
  sc: SeedCtx,
  md: MasterData,
): Promise<void> {
  const transfer = await deps.stockTransfers.create(
    permitFor(deps, actor, 'invStockTransfers', 'create'),
    {
      docDate: daysAgo(10),
      summary: '成品转仓调拨',
      remarks: '初始化示例调拨单',
      companyId: sc.company.id,
      fromWarehouseId: sc.warehouses.default,
      toWarehouseId: sc.warehouses.finished,
      transitWarehouseId: sc.warehouses.transit,
    },
  )
  const lines = [
    { key: 'box_shell', qty: 15 },
    { key: 'terminal_block', qty: 400 },
  ]
  const itemIds: string[] = []
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i]!
    const mat = md.materials[line.key]!
    const item = await deps.stockTransfers.createItem(
      permitFor(deps, actor, 'invStockTransferItems', 'create'),
      {
        stockTransferId: transfer.id,
        idx: i + 1,
        qty: String(line.qty),
        materialId: mat.id,
        unitId: mat.defaultUnitId,
      },
    )
    itemIds.push(item.id)
  }
  await deps.stockTransfers.ship(permitFor(deps, actor, 'invStockTransfers', 'ship'), transfer.id)
  await deps.stockTransfers.receive(
    permitFor(deps, actor, 'invStockTransfers', 'receive'),
    transfer.id,
    {
      receipts: [
        { itemId: itemIds[0]!, qty: '15' },
        { itemId: itemIds[1]!, qty: '250' },
      ],
    },
  )
}

async function seedStockCount(
  deps: SampleDataDeps,
  actor: Actor,
  sc: SeedCtx,
  md: MasterData,
): Promise<void> {
  const count = await deps.stockCounts.create(permitFor(deps, actor, 'invStockCounts', 'create'), {
    postingDate: daysAgo(3),
    summary: '月末例行盘点',
    remarks: '初始化示例盘点单',
    companyId: sc.company.id,
    warehouseId: sc.warehouses.default,
    loadAll: true,
  })
  const listed = await deps.stockCounts.queryItems(
    permitFor(deps, actor, 'invStockCountItems', 'read'),
    {
      filter: { countId: { kind: 'fk', values: [count.id], labels: [] } },
      limit: 200,
    },
  )
  const items = listed.results
  if (items.length === 0) {
    throw new ApiError('conflict', '示例数据盘点整仓带出失败:默认仓库无账面余额')
  }
  const screwId = md.materials.screw!.id
  const railId = md.materials.rail!.id
  for (const item of items) {
    const book = decimal(item.bookQuantity)
    let counted = book
    if (item.materialId === screwId) counted = book.sub(50)
    if (item.materialId === railId) counted = book.add(5)
    await deps.stockCounts.updateItem(
      permitFor(deps, actor, 'invStockCountItems', 'update'),
      item.id,
      { countedQuantity: counted.toFixed(), countedQuantityPresent: true },
    )
  }
  await deps.stockCounts.approve(permitFor(deps, actor, 'invStockCounts', 'approve'), count.id)
}
